packages/clerk-js/src/utils/captcha/turnstile.ts

```diff
@@ -112,59 +116,71 @@ async function loadCaptchaFromFAPIProxiedURL(fallbackUrl: string) {
  * - If the widgetType is 'smart', the captcha widget is rendered in a div with the id 'clerk-captcha'. If the div does
  *  not exist, the invisibleSiteKey is used as a fallback and the widget is rendered in a hidden div at the bottom of the body.
  */
-export const getTunstileToken = async (captchaOptions: {
-  siteKey: string;
-  scriptUrl: string;
-  widgetType: CaptchaWidgetType;
-  invisibleSiteKey: string;
-}) => {
-  const { siteKey, scriptUrl, widgetType, invisibleSiteKey } = captchaOptions;
-  let captchaToken = '',
-    id = '';
-  let isInvisibleWidget = !widgetType || widgetType === 'invisible';
+export const getTurnstileToken = async (opts: CaptchaOptions) => {
+  const { siteKey, scriptUrl, widgetType, invisibleSiteKey } = opts;
+  const { modalContainerQuerySelector, modalWrapperQuerySelector, closeModal, openModal } = opts;
+  const captcha: Turnstile = await loadCaptcha(scriptUrl);
+  const errorCodes: (string | number)[] = [];
+
+  let captchaToken = '';
+  let id = '';
   let turnstileSiteKey = siteKey;
+  let retries = 0;
+  let widgetContainerQuerySelector: string | undefined;
+  const widgetTypeUsed = [widgetType || ''];
+
+  // modal
+  if (modalContainerQuerySelector && modalWrapperQuerySelector) {
+    widgetContainerQuerySelector = modalContainerQuerySelector;
+    await openModal?.();
+    await waitForElement(modalContainerQuerySelector);
+  }
 
-  let widgetDiv: HTMLElement | null = null;
+  // smart widget with container provided by user
+  if (!widgetContainerQuerySelector && widgetType === 'smart') {
+    widgetContainerQuerySelector = `#${CAPTCHA_ELEMENT_ID}`;
+    const visibleDiv = document.getElementById(CAPTCHA_ELEMENT_ID);
+    if (visibleDiv) {
+      visibleDiv.style.display = 'block';
+    } else {
+      console.error(
+        'Cannot initialize Smart CAPTCHA widget because the `clerk-captcha` DOM element was not found; falling back to Invisible CAPTCHA widget. If you are using a custom flow, visit https://clerk.com/docs/custom-flows/bot-sign-up-protection for instructions',
+      );
+    }
+  }
 
-  const createInvisibleDOMElement = () => {
+  // invisible widget for which we create the container automatically
+  if (!widgetContainerQuerySelector) {
+    turnstileSiteKey = invisibleSiteKey;
+    widgetTypeUsed.push('invisible');
```

Comment:
Again, here the `widgetTypeUsed` will end up having `['invisible','invisible']` as a value, if the user has selected the `Invisible` option from the Dashboard.